File: 3.Longest_Substring_Without_Repeating_Characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        if(s == ""){
            return 0;
        }
        int n=s.length();
        int cur_length = 1;
        int max_length = 1;
        int prev_index;
        int i;
        int visited[256];
        bool flag = false;

        for(i=0;i<256;i++){
            visited[i]=-1;
        }

        visited[s[0]] = 0;

        for(i=1;i<n;i++){
            prev_index = visited[s[i]];
            if(prev_index == -1 || i-cur_length > prev_index){
                cur_length++;
            }
            else{
                if(flag==false)
                    flag = true;
                if(cur_length > max_length){
                    max_length = cur_length;
                }
                cur_length =  i - prev_index;
            }
            visited[s[i]] = i;
        }
        if(cur_length > max_length){
                    max_length = cur_length;
        }
        return max_length;
        //if(flag)
            //return cur_length;
        //else
            //return max_length;
    }
};

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        string_map={}
        left=0
        right=0
        max_length=0
        while right<len(s) and left<len(s):
            if s[right] in string_map:
                left=max(left,string_map[s[right]]+1)
            string_map[s[right]]=right
            max_length=max(max_length, right-left+1)
            right+=1
        return max_length